distributed/scheduler.py

```diff
@@ -2422,61 +2830,59 @@ def transition_waiting_processing(self, key):
 
     def transition_waiting_memory(self, key, nbytes=None, worker=None, **kwargs):
         try:
+            ws = self.workers[worker]
+            ts = self.task_states[key]
+
             if self.validate:
-                assert key not in self.rprocessing
-                assert key in self.waiting
-                assert self.task_state[key] == 'waiting'
+                assert not ts.processing_on
+                assert ts.waiting_on
+                assert ts.state == 'waiting'
 
-            del self.waiting[key]
+            ts.waiting_on.clear()
 
             if nbytes is not None:
-                self.nbytes[key] = nbytes
-
-            if key not in self.who_has:
-                self.who_has[key] = set()
+                ts.nbytes = nbytes
 
-            self.who_has[key].add(worker)
-            self.has_what[worker].add(key)
-            self.worker_bytes[worker] += nbytes or DEFAULT_DATA_SIZE
+            ts.who_has.add(ws)
+            ws.has_what.add(ts)
+            ws.nbytes += ts.get_nbytes()
 
-            self.check_idle_saturated(worker)
+            self.check_idle_saturated(ws)
 
             recommendations = OrderedDict()
 
-            deps = self.dependents.get(key, [])
+            # XXX factor this out? (see transition_processing_memory)
+            deps = ts.dependents
             if len(deps) > 1:
-                deps = sorted(deps, key=self.priority.get, reverse=True)
-
-            for dep in deps:
-                if dep in self.waiting:
-                    s = self.waiting[dep]
-                    s.remove(key)
+                deps = sorted(deps, key=operator.attrgetter('priority'),
+                              reverse=True)
+            for dts in deps:
+                s = dts.waiting_on
+                if ts in s:
+                    s.discard(ts)
                     if not s:  # new task ready to run
-                        recommendations[dep] = 'processing'
-
-            for dep in self.dependencies.get(key, []):
-                if dep in self.waiting_data:
-                    s = self.waiting_data[dep]
-                    s.remove(key)
-                    if (not s and dep and
-                        dep not in self.who_wants and
-                            not self.waiting_data.get(dep)):
-                        recommendations[dep] = 'released'
-
-            if (not self.waiting_data.get(key) and
-                    key not in self.who_wants):
+                        recommendations[dts.key] = 'processing'
+
+            for dts in ts.dependencies:
+                s = dts.waiters
+                s.discard(ts)
+                if not s and not dts.who_wants:
+                    recommendations[dts.key] = 'released'
+
+            if not ts.waiters and not ts.who_wants:
+                # XXX what about 'fire-and-forget'?
```

Comment:
Fire-and-forget should just be another client in `ts.who_wants`, no?

Reply:
Well, yes, but fire-and-forget doesn't require the task to remain alive.